yarn-project/p2p/src/services/tx_collection/tx_collection.ts

```diff
@@ -0,0 +1,280 @@
+  public collectFastForProposal(
+    blockProposal: BlockProposal,
+    txHashes: TxHash[] | string[],
+    opts: { deadline: Date; pinnedPeer?: PeerId },
+  ) {
+    return this.collectFastFor({ type: 'proposal', blockProposal }, txHashes, opts);
+  }
+
+  /** Collects the set of txs for the given mined block as fast as possible */
+  public collectFastForBlock(
+    block: L2Block,
+    txHashes: TxHash[] | string[],
+    opts: { deadline: Date; pinnedPeer?: PeerId },
+  ) {
+    return this.collectFastFor({ type: 'block', block }, txHashes, opts);
+  }
+
+  /** Collects the set of txs for the given proposal or block as fast as possible */
+  public collectFastFor(
+    input: FastCollectionRequestInput,
+    txHashes: TxHash[] | string[],
+    opts: { deadline: Date; pinnedPeer?: PeerId },
+  ) {
+    return this.fastCollection.collectFastFor(input, txHashes, opts);
+  }
+
+  /**
+   * Wrapper for a collection function. Handles logging, metrics, removing found txs from the missing set,
+   * adding them to the found set (for fast requests only), and adding to the tx pool. Called by both fast and slow
+   * collection methods whenever they actually collect txs.
+   */
+  private async collect(
+    collectFn: (txHashes: TxHash[]) => Promise<(Tx | undefined)[]>,
+    requested: TxHash[],
+    info: Record<string, any> & { description: string; method: CollectionMethod },
+  ) {
+    this.log.trace(`Requesting ${requested.length} txs via ${info.description}`, {
+      ...info,
+      requestedTxs: requested.map(t => t.toString()),
+    });
+
+    // Execute collection function and measure the time taken, catching any errors.
+    const [duration, txs] = await elapsed(async () => {
+      try {
+        const response = await collectFn(requested);
+        return await Tx.toTxsWithHashes(response.filter(tx => tx !== undefined));
+      } catch (err) {
+        this.log.error(`Error collecting txs via ${info.description}`, err, {
+          ...info,
+          requestedTxs: requested.map(hash => hash.toString()),
+        });
+        return [] as TxWithHash[];
+      }
+    });
+
+    if (txs.length === 0) {
+      this.log.trace(`No txs found via ${info.description}`, {
+        ...info,
+        requestedTxs: requested.map(t => t.toString()),
+      });
+      return { txs, requested, duration };
+    }
+
+    this.log.verbose(
+      `Collected ${txs.length} txs out of ${requested.length} requested via ${info.description} in ${duration}ms`,
+      { ...info, duration, txs: txs.map(t => t.txHash.toString()), requestedTxs: requested.map(t => t.toString()) },
+    );
+
+    // Report metrics for the collection
+    this.instrumentation.increaseTxsFor(info.method, txs.length, duration);
+
+    // Mark txs as found in the slow missing txs set and all fast requests
+    this.foundTxs(txs);
+
+    // Add the txs to the tx pool (should not fail, but we catch it just in case)
+    try {
+      await this.txPool.addTxs(txs, { source: `tx-collection` });
```

Comment:
there's an implicit trust that `collectFn` validates these txs. Maybe it's worth renaming to `collectValidTxs`?